Ghosts cant mess with RCD & Plumbing RCD

SyncIt21 opened the pull request:
## About The Pull Request
Fixes #72036

## Changelog
:cl:
fix: ghosts cant mess with RCD & PLumbing RCD
/:cl:

SyncIt21 committed:
Ghosts can mess with this

SyncIt21 commented:
Wrong description for the commit . What i mean is they can't/cannot mess with it

SyncIt21 committed:
Merge branch 'master' into Fix

SyncIt21 commented:
@Mothblocks

Mothblocks commented:
this really is not the right solution. when you are returning `always_state`, that is where you are going wrong. `canUseTopic` shouldn't exist in any tgui proc like this

Mothblocks commented:
`always_state` is for magical UIs not for items like this. you are doing it wrong

SyncIt21 commented:
> `canUseTopic` shouldn't exist in any tgui proc like this

It exists in RPD `ui_act()` and it has been there for a long time

> always_state is for magical UIs not for items like this. you are doing it wrong

Like I said the window won't open in the `base_construction` console without this line so if you have a better suggestion I'm open to it.

Mothblocks commented:
> It exists in RPD ui_act() and it has been there for a long time

It should not be, ui_status() should be its replacement.

> Like I said the window won't open in the base_construction console without this line so if you have a better suggestion I'm open to it.

Ok, first explain why that's the case and why `ui_host()` doesn't fix it.

Second, if there's a good reason, what you can do is return always_state, but override `/ui_status` to define more precise checks on what *specific mobs* are allowed to do. This is what the strip menu does, for instance, since it needs more precise filtering.

There's a bunch of other states as well, I need you to explain what *specifically* the issue is that you are trying to solve here

SyncIt21 commented:
So the previous code inside construction cosole actions used an `tgui_input_list()` to select the build mode and `tgui_input_list()` returns always active for its ui_state and thats where I got this idea.

I'm not trying to do anything specific really I just want the window to open up in `base_construction` console and now since your suggesting to use ui_state() in the RPD as well then what should be return values in both these cases?

I'm new to using both these procs so some guidance would be really useful

Mothblocks commented:
> As to why ui_host() doesn't work it didn't prevent ghosts from tampering with the RCD. 

`ui_host()` just specifies what "thing" represents the UI. I'm assuming you want something like that here. Ghosts won't be able to mess with it if you don't have it set to `always_state`, which is only supposed to be for either OOC non-in-world UIs or for implementing `ui_status()`.

Where is the construction console? What does not setting `ui_state()` do that you want to avoid?

SyncIt21 commented:
Ok now im also confused. Ill start from the beginning

> Where is the construction console?

```
obj/machinery/computer/camera_advanced/base_construction/aux[for player]
obj/machinery/computer/camera_advanced/base_construction/centcom[for admin]
```
When you open this console you get ui actions on the top left corner of your screen. And i am talking specifically about the circled RCD icon
![Screenshot (68)](https://user-images.githubusercontent.com/110812394/208360947-562d1b19-ddd7-4087-bde2-025242aab48c.png)
If i don't do this
```
/obj/item/construction/rcd/ui_state(mob/user) 
 	return GLOB.always_state
```
Then the RCD Window does not show up when i click this icon and this is the only reason ever why I'm using this code. Now what you are suggesting is replacing this `ui_state() `with `ui_host()` to get the same results but what I'm saying is it didn't work. I dont understand how these 2 methods are even related or how one can replace the other. 

Now if you are suggesting that i use `ui_host()` to replace `usr.canUseTopic()` That i 100% agree with you I'm trying to do that right now but if you are suggesting `ui_host()` to replace` ui_state()` that i don't understand how that works because i did try to do that and the Window does not open.

So the question i have is are you suggesting to use `ui_host()` to replace both `usr.canUseTopic()` & ` ui_state()` or are you suggestign that `ui_host()` should only replace `usr.canUseTopic()`? 

Also my head hurts cause i dont know if i explained any of this correctly :(

Mothblocks commented:
What you want to do is, instead of having the "always active" var, you want to provide a var that says "use this as the host", which `ui_host()` will return, ex:

```dm
/obj/item/rcd/ui_host(mob/user)
    return explicit_ui_host || ..() // The console sets explicit_ui_host
```

You shouldn't need canUseTopic or any of that. See how far this gets you.

SyncIt21 commented:
> What you want to do is, instead of having the "always active" var, you want to provide a var that says "use this as the host", which `ui_host()` will return, ex:
> 
> ```dm
> /obj/item/rcd/ui_host(mob/user)
>     return explicit_ui_host || ..() // The console sets explicit_ui_host
> ```
> 
> You shouldn't need canUseTopic or any of that. See how far this gets you.

Ok good news your suggestion worked i.e. i can now open the window inside the construction console without `ui_state()`

Bad news is ghosts can still mess the hand held RCD since i removed `usr.canUseTopic()`

Any ideas?

Mothblocks commented:
That happens without setting ui_state at all? Can you push your code?

SyncIt21 commented:
> That happens without setting ui_state at all? Can you push your code?

Yes i removed `ui_state()` &` usr.canUseTopic()` and replaced both with your suggestion. Pushing the code now

SyncIt21 committed:
ui_host()

Mothblocks commented:
That, definitely shouldn't be happening O_o

Default UI state automatically closes for non-livings. I'll test your branch locally tomorrow-ish

Are you sure it's not because you were adminned?

SyncIt21 commented:
> That, definitely shouldn't be happening O_o
> 
> Default UI state automatically closes for non-livings. I'll test your branch locally tomorrow-ish
> 
> Are you sure it's not because you were adminned? (And of course, though it sounds silly, make sure you recompiled)

Ok ill de-admin myself & try it. Give me a minute

SyncIt21 commented:
😭 it didnt work. Both admin & non admins ghosts can mess with this

Mothblocks commented:
I'll give it a look later then, can you give me repro steps?

SyncIt21 commented:
1. Create/Spawn an RCD
2. Turn yourself into a ghost either with admin tools or you know suicide
3. click on RCD & change anything you want there

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -643,8 +637,6 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	..()
```
Oh!

It's because this is `..()` and not `. = ..()`

Meaning you never actually get the fail value

(You might not even need ui_host() at all, give that a try)

Mothblocks replied:
Make sure RPDs do this too

SyncIt21 replied:
oh my god. give me a sec

Mothblocks replied:
This would be awesome to lint for but SpacemanDMM doesn't give us that capability yet, wonder what else doesn't have this

SyncIt21 replied:
Nope it didnt work :(. You definitly caught a bug thanks for that but ghosts can still have their fun with the RCD and the console still needs `ui_host()` to open

It also makes sense why it does not work because if you look inside RPD` ui_act()`
```
. = ..()
	if(.)
		return
	if(!usr.canUseTopic(src, be_close = TRUE))
		return
```
You will they didnt make the same mistake as me and still used `canUseTopic()`

Im going no where fast

SyncIt21 committed:
Dot Dot Dot

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -1237,6 +1231,8 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	. = ..()
 	if(.)
 		return
+	if(!usr.canUseTopic(src, be_close = TRUE))
```
You should now be able to safely remove this

SyncIt21 replied:
Um did you read the comment in your last review?

> Nope it didnt work :(. You definitly caught a bug thanks for that but ghosts can still have their fun with the RCD and the console still needs ui_host() to open

RCD & Plumbing RCD still requires canUseTopic() else ghosts can still mess with it. I resolved that conversation because i made the change you asked me too but not because it solved this problem

Mothblocks replied:
Going to take a look at this again tomorrow then because this is 100% not necessary, nothing else needs this

SyncIt21 commented:
After 4 hours of recompiling, Typing Dots & several sessions of physical therapy later. I am proud to say
Its finally working :)

Just needed a few more dots

`ui_host()` is still required for the console to open though, and you were right turns out pipe_dispenser never needed `canUseTopic()` dont know why they put it there

SyncIt21 committed:
YES FUCKING FINALLY

SyncIt21 commented:
Can we end this? I need a drink

Mothblocks commented:
https://youtu.be/RlLl6bBq584?t=111

SyncIt21 commented:
We seriously need a linter for this but for now I'm heading out
